### ISSUE TITLE
Compression JSON output

### ISSUE THREAD

**Issue by gopalakrishnan-subramani** [strongloop/loopback-next#3277]
We have huge JSON content to be send to client, with compression, the size reduced to 1/10.

We could do easily in express framework, but I could not figure the same in Loopback 4. Seems like Sequence is a way to go, but struggling. Can you give a documentation ref or example or idea to implement the same?

https://github.com/expressjs/compression

**Comment by joeysino** [comment]
It's possible to get [the express app](https://github.com/strongloop/loopback-next/blob/2118d806f22aa66825b9b25348ae41de5f30952f/packages/rest/src/rest.server.ts#L220) from the RestApplication's RestServer.

But when I added middleware, it seemed it was never being called.

This is what I tried, in various places in the constructor of `application.ts`:

```ts
  const expressApp = this.requestHandler as express.Application;
  if (!expressApp.use) throw new Error("RestApplication's requestHandler is not express!");
  expressApp.use(compression());
```

I'm running an older version: `@loopback/rest@^1.1.0`

**Comment by joeytwiddle** [comment]
It's possible to get [the express app](https://github.com/strongloop/loopback-next/blob/2118d806f22aa66825b9b25348ae41de5f30952f/packages/rest/src/rest.server.ts#L220) from the RestApplication's RestServer.

But when I added middleware, it seemed it was never being called.

This is what I tried, in various places in the constructor of `application.ts`:

```ts
  const expressApp = this.requestHandler as express.Application;
  if (!expressApp.use) throw new Error("RestApplication's requestHandler is not express!");
  expressApp.use(compression());
```

I'm running an older version: `@loopback/rest@^1.1.0`

**Comment by bajtos** [comment]
@gopalakrishnan-subramani thank you for reporting this issue. 

At the moment, we don't have a first-class support for Express middleware in LoopBack 4. It's a missing feature we are aware of, see the discussion in https://github.com/strongloop/loopback-next/issues/1293 for more details.

Our current recommendation is to create a top-level Express application (where you can easily mount any middleware) and then mount LB4 as a component (middleware) in that application. See https://loopback.io/doc/en/lb4/express-with-lb4-rest-tutorial.html for detailed instructions.

I don't think we will be implementing custom compression implementation (e.g. a custom Sequence action), IMO we should always advise developers to use Express middleware `compression`.

Let's close this feature as a duplicate of #1293 

---

@joeytwiddle 
> It's possible to get the express app from the RestApplication's RestServer.
> But when I added middleware, it seemed it was never being called.

I think you are adding `compression` middleware too late in the configuration process. It's added to the end of the middleware chain, it would be invoked only after all previously registered middleware (like LB4 router) did not handle the request and called `next()`. LB4 router never (*) calls `next`, therefore your middleware is never invoked.

(*) If the promise returned by the Sequence is rejected, we call `next(err)`. However, this happens only when something goes terribly wrong. In most cases, errors are handled by Sequence action `reject` which sends back HTTP error response and thus consider the error as handled.
